File: acs_matcher/match_participants.py
import os
import re
import hashlib

import numpy as np
import pandas as pd

"""
Core matching logic for mapping participant-level rows to ACS tracts and
copying over a `new_feature` column.
"""

# --- helpers (private to the module) ---

_num_re = re.compile(r"[-+]?\d*\.?\d+(?:[eE][-+]?\d+)?")


def _alias_for_geoid(geoid: str) -> str:
    """Deterministic alias for a GEOID, to avoid exposing the raw tract ID."""
    h = hashlib.sha1(str(geoid).encode("utf-8")).hexdigest()[:8]
    return f"alias_{h}"


def _to_numeric_series(s: pd.Series) -> pd.Series:
    """
    Convert a Series to numeric where possible, robust to commas and mixed text.
    Non-parsable values become NaN.
    """
    if pd.api.types.is_numeric_dtype(s):
        return pd.to_numeric(s, errors="coerce")

    def parse_cell(x):
        if pd.isna(x):
            return np.nan
        x = str(x).replace(",", "")
        m = _num_re.search(x)
        return float(m.group(0)) if m else np.nan

    return s.map(parse_cell)


def match_participants(
    acs_csv_path: str,
    participant_csv_path: str,
    rtol: float = 0.005,
):
    """
    Match participant-level rows to ACS tracts and copy over `new_feature`. It essentially matches participants to a new feature.

    Parameters
    ----------
    acs_csv_path : str
        Path to ACS CSV with columns:
        - 'geoid'
        - neighborhood feature columns
        - 'new_feature' (the column to transfer)
    participant_csv_path : str
        Path to participant CSV with columns:
        - 'id'
        - any subset of the same neighborhood features as ACS
    rtol : float, optional
        Relative tolerance for matching (default 0.005 = 0.5%).

    Matching logic
    --------------
    - Columns in the participant file that are not present in ACS are ignored
      (a warning is printed).
    - Requires:
        * participant CSV has 'id'
        * ACS CSV has 'geoid' and 'new_feature'
    - For each participant row:
        * consider all overlapping feature columns
        * ignore features that are NaN for that participant
        * restrict ACS comparison to those available features
        * compute relative differences; require ALL used features to be within `rtol`
        * if multiple tracts match, pick the one with minimum mean relative diff.

    Outputs
    -------
    Two CSVs are written next to the participant file:

        <participant_input>_matched.csv
            - original participant columns
            - plus: 'new_feature', 'tract_alias'

        <participant_input>_unmatched.csv
            - original participant rows that found no match (no extra cols)

    Returns
    -------
    matched_path : str
        Path to the matched CSV.
    unmatched_path : str
        Path to the unmatched CSV.
    """
    # --- load and normalize column names ---
    acs_df_raw = pd.read_csv(acs_csv_path, dtype=str)
    acs_df_raw.columns = [c.strip() for c in acs_df_raw.columns]

    user_df_raw = pd.read_csv(participant_csv_path, dtype=str)
    user_df_raw.columns = [c.strip() for c in user_df_raw.columns]

    # --- required columns ---
    if "id" not in user_df_raw.columns:
        raise ValueError("Participant file must contain an 'id' column.")
    if "geoid" not in acs_df_raw.columns:
        raise ValueError("ACS data file must contain a 'geoid' column.")
    if "new_feature" not in acs_df_raw.columns:
        raise ValueError("ACS data file must contain a 'new_feature' column.")

    # --- features and overlap ---
    user_feature_cols = [c for c in user_df_raw.columns if c != "id"]
    acs_feature_cols = [c for c in acs_df_raw.columns if c not in ("geoid", "new_feature")]

    extra = [c for c in user_feature_cols if c not in acs_feature_cols]
    if extra:
        print("Columns in participant CSV not found in ACS Data (ignored):")
        for c in extra:
            print("  -", c)

    overlap = [c for c in user_feature_cols if c in acs_feature_cols]
    if not overlap:
        raise ValueError("No overlapping neighborhood feature columns between participant CSV and ACS.")

    # --- numeric versions for comparison (outputs keep original values) ---
    user_cmp = pd.DataFrame({c: _to_numeric_series(user_df_raw[c]) for c in overlap})
    acs_cmp = pd.DataFrame({c: _to_numeric_series(acs_df_raw[c]) for c in overlap})

    # ensure there is at least one participant row with some numeric data
    if (~user_cmp[overlap].isna().all(axis=1)).sum() == 0:
        raise ValueError("All participant rows are empty/NaN across the overlapping columns.")

    # --- precalc ACS arrays for speed ---
    acs_vals = acs_cmp.to_numpy(dtype=float)  # (M, K)
    acs_new = pd.to_numeric(acs_df_raw["new_feature"], errors="coerce").to_numpy()
    acs_alias = acs_df_raw["geoid"].map(_alias_for_geoid).to_numpy()

    eps = 1e-12
    matched_rows, unmatched_rows = [], []

    for idx, row in user_df_raw.iterrows():
        row_cmp = user_cmp.iloc[idx]
        valid_cols = [j for j, c in enumerate(overlap) if pd.notna(row_cmp[c])]
        if not valid_cols:
            # no usable feature values for this participant
            unmatched_rows.append(row.to_dict())
            continue

        # restrict to columns with values for this participant
        uv = row_cmp[overlap].to_numpy(dtype=float)[valid_cols]
        av = acs_vals[:, valid_cols]

        # relative diffs
        denom = np.maximum(np.maximum(np.abs(av), np.abs(uv)), eps)
        rel = np.abs(av - uv) / denom

        # require ALL provided features within tolerance
        within = (rel <= rtol).all(axis=1)

        if within.any():
            cand_rel = rel[within]
            best_idx_within = int(np.argmin(cand_rel.mean(axis=1)))
            best_global_idx = np.flatnonzero(within)[best_idx_within]

            matched_rows.append(
                {
                    **row.to_dict(),
                    "new_feature": acs_new[best_global_idx],
                    "tract_alias": acs_alias[best_global_idx],
                }
            )
        else:
            unmatched_rows.append(row.to_dict())

    # --- build outputs (keep original participant columns order) ---
    matched_df = pd.DataFrame(
        matched_rows,
        columns=list(user_df_raw.columns) + ["new_feature", "tract_alias"],
    )
    unmatched_df = pd.DataFrame(unmatched_rows, columns=list(user_df_raw.columns))

    # --- write next to participant file ---
    base, _ = os.path.splitext(participant_csv_path)
    matched_path = f"{base}_matched.csv"
    unmatched_path = f"{base}_unmatched.csv"
    matched_df.to_csv(matched_path, index=False)
    unmatched_df.to_csv(unmatched_path, index=False)

    print(f"Matched: {len(matched_df):,}. Unmatched: {len(unmatched_df):,}")
    print(f" - {matched_path}")
    print(f" - {unmatched_path}")

    return matched_path, unmatched_path

File: version-1/r/R/match_participants.R
# Match participant rows to ACS tracts and copy over `new_feature`.
# Mirrors the Python implementation in `acs_matcher.match_participants`.

# NOTE: Requires the 'digest' package for SHA1.
# install.packages("digest")

# Deterministic alias for a GEOID (matches Python hashlib.sha1(str(geoid).encode("utf-8")).hexdigest()[:8])
.alias_for_geoid <- function(geoid) {
  s <- enc2utf8(as.character(geoid))
  h <- digest::digest(s, algo = "sha1", serialize = FALSE)
  paste0("alias_", substr(h, 1, 8))
}

# Convert a vector to numeric where possible, robust to commas and mixed text.
# Non-parsable values become NA.
.to_numeric_series <- function(x) {
  if (is.numeric(x)) {
    return(suppressWarnings(as.numeric(x)))
  }
  
  parse_cell <- function(val) {
    if (is.na(val)) {
      return(NA_real_)
    }
    cleaned <- gsub(",", "", as.character(val), perl = TRUE)
    m <- regexpr("[-+]?\\d*\\.?\\d+(?:[eE][-+]?\\d+)?", cleaned, perl = TRUE)
    if (m[1] == -1) {
      return(NA_real_)
    }
    as.numeric(substr(cleaned, m[1], m[1] + attr(m, "match.length") - 1))
  }
  
  vapply(x, parse_cell, numeric(1))
}

# Match participant rows to ACS tracts and copy over `new_feature`.
#This mirrors the Python implementation in `acs_matcher.match_participants`.

#' @param acs_csv_path Path to ACS CSV with columns: geoid, neighborhood features, new_feature
#' @param participant_csv_path Path to participant CSV with columns: id, subset of neighborhood features
#' @param rtol Relative tolerance for matching (default 0.005 = 0.5%)

#' @return A named list with matched_path and unmatched_path
match_participants <- function(acs_csv_path, participant_csv_path, rtol = 0.005) {
  eps <- 1e-12
  
  # Read everything as character to mirror Python pd.read_csv(..., dtype=str)
  acs_df <- utils::read.csv(acs_csv_path, colClasses = "character", check.names = FALSE)
  user_df <- utils::read.csv(participant_csv_path, colClasses = "character", check.names = FALSE)
  
  names(acs_df) <- trimws(names(acs_df))
  names(user_df) <- trimws(names(user_df))
  
  # Required columns
  if (!("id" %in% names(user_df))) {
    stop("Participant file must contain an 'id' column.")
  }
  if (!("geoid" %in% names(acs_df))) {
    stop("ACS data file must contain a 'geoid' column.")
  }
  if (!("new_feature" %in% names(acs_df))) {
    stop("ACS data file must contain a 'new_feature' column.")
  }
  
  # Features and overlap
  user_feature_cols <- setdiff(names(user_df), "id")
  acs_feature_cols <- setdiff(names(acs_df), c("geoid", "new_feature"))
  
  extra <- setdiff(user_feature_cols, acs_feature_cols)
  if (length(extra) > 0) {
    warning(
      "Columns in participant CSV not found in ACS Data (ignored): ",
      paste(extra, collapse = ", ")
    )
  }
  
  overlap <- intersect(user_feature_cols, acs_feature_cols)
  if (length(overlap) == 0) {
    stop("No overlapping neighborhood feature columns between participant CSV and ACS.")
  }
  
  # Numeric versions for comparison (outputs keep original character values)
  user_cmp <- data.frame(lapply(user_df[overlap], .to_numeric_series), check.names = FALSE)
  acs_cmp  <- data.frame(lapply(acs_df[overlap], .to_numeric_series), check.names = FALSE)
  
  # Ensure at least one participant row has some numeric data
  if (sum(rowSums(!is.na(user_cmp)) > 0) == 0) {
    stop("All participant rows are empty/NaN across the overlapping columns.")
  }
  
  # Precompute ACS arrays
  acs_vals  <- as.matrix(acs_cmp)  # numeric matrix (M, K)
  storage.mode(acs_vals) <- "double"
  
  acs_new   <- suppressWarnings(as.numeric(acs_df[["new_feature"]]))
  acs_alias <- vapply(acs_df[["geoid"]], .alias_for_geoid, character(1), USE.NAMES = FALSE)
  
  matched_rows <- list()
  unmatched_rows <- list()
  
  for (i in seq_len(nrow(user_df))) {
    row_cmp <- user_cmp[i, , drop = FALSE]          # 1-row data.frame of numeric comparisons
    valid_cols <- which(!is.na(row_cmp[1, ]))       # indices of overlap cols with values for this participant
    
    if (length(valid_cols) == 0) {
      unmatched_rows[[length(unmatched_rows) + 1]] <- user_df[i, , drop = FALSE]
      next
    }
    
    # Participant vector and ACS matrix restricted to valid columns
    uv <- as.numeric(row_cmp[1, valid_cols, drop = TRUE])  # numeric vector length = K_valid
    av <- acs_vals[, valid_cols, drop = FALSE]             # (M, K_valid)
    
    # Relative diffs: denom = max(max(|av|, |uv|), eps)
    uv_mat <- matrix(uv, nrow = nrow(av), ncol = length(uv), byrow = TRUE)
    denom <- pmax(pmax(abs(av), abs(uv_mat)), eps)
    rel <- abs(av - uv_mat) / denom
    
    # Require ALL provided features within tolerance
    within <- apply(rel <= rtol, 1, all)
    
    if (any(within)) {
      cand_rel <- rel[within, , drop = FALSE]
      mean_rel <- rowMeans(cand_rel)
      best_idx_within <- which.min(mean_rel)
      best_global_idx <- which(within)[best_idx_within]
      
      out_row <- cbind(
        user_df[i, , drop = FALSE],
        data.frame(
          new_feature = acs_new[best_global_idx],
          tract_alias = acs_alias[best_global_idx],
          check.names = FALSE,
          stringsAsFactors = FALSE
        )
      )
      matched_rows[[length(matched_rows) + 1]] <- out_row
    } else {
      unmatched_rows[[length(unmatched_rows) + 1]] <- user_df[i, , drop = FALSE]
    }
  }
  
  # Build outputs (keep original participant column order)
  matched_cols <- c(names(user_df), "new_feature", "tract_alias")
  matched_df <- if (length(matched_rows) > 0) {
    df <- do.call(rbind, matched_rows)
    df <- df[, matched_cols, drop = FALSE]
    df
  } else {
    stats::setNames(data.frame(matrix(ncol = length(matched_cols), nrow = 0), check.names = FALSE), matched_cols)
  }
  
  unmatched_df <- if (length(unmatched_rows) > 0) {
    do.call(rbind, unmatched_rows)
  } else {
    user_df[0, , drop = FALSE]
  }
  
  # Write next to participant file
  base <- tools::file_path_sans_ext(participant_csv_path)
  matched_path <- paste0(base, "_matched.csv")
  unmatched_path <- paste0(base, "_unmatched.csv")
  
  utils::write.csv(matched_df, matched_path, row.names = FALSE, quote = TRUE)
  utils::write.csv(unmatched_df, unmatched_path, row.names = FALSE, quote = TRUE)
  
  message("Matched: ", nrow(matched_df), ". Unmatched: ", nrow(unmatched_df))
  message(" - ", matched_path)
  message(" - ", unmatched_path)
  
  list(matched_path = matched_path, unmatched_path = unmatched_path)
}
